File: examples/bench/fp_dot_compare.py
#!/usr/bin/env python3

def make_x(n: int):
    return [i * 0.5 + 1.1 for i in range(n)]


def make_y(n: int):
    return [i * 0.25 + 2.3 for i in range(n)]


def dot(x, y) -> float:
    acc = 0.0
    for a, b in zip(x, y):
        acc += a * b
    return acc


if __name__ == "__main__":
    n = 2000
    rounds = 200
    x = make_x(n)
    y = make_y(n)
    total = 0.0
    for _ in range(rounds):
        total += dot(x, y)
    print(f"{total:.12f}")

File: benchmarks/bench/fp_dot_compare.c
#include <stdio.h>
#include <stdlib.h>

static double dot(const double *x, const double *y, int n) {
    double acc = 0.0;
    for (int i = 0; i < n; i++) {
        acc += x[i] * y[i];
    }
    return acc;
}

int main(void) {
    const int n = 2000;
    const int rounds = 200;

    double *x = (double *)malloc((size_t)n * sizeof(double));
    double *y = (double *)malloc((size_t)n * sizeof(double));
    if (!x || !y) {
        free(x);
        free(y);
        return 1;
    }

    for (int i = 0; i < n; i++) {
        x[i] = (double)i * 0.5 + 1.1;
        y[i] = (double)i * 0.25 + 2.3;
    }

    double total = 0.0;
    for (int r = 0; r < rounds; r++) {
        total += dot(x, y, n);
    }

    printf("%.12f\n", total);
    free(x);
    free(y);
    return 0;
}
